### PR TITLE
Introduce internal data model for GTI

### PR DESCRIPTION
**Pull request by registerrier**
**Description**

This pull request modifies `GTI` to rely on a `Table` object that does not follow GADF. Instead, it uses start and stop `Time` columns. The `read` and `write` methods so far support only gadf.

**Dear reviewer**

### CODE REVIEW

File: gammapy/data/gti.py [hunk 1]
```diff
@@ -78,21 +96,20 @@ def create(cls, start, stop, reference_time="2000-01-01"):
...
-        meta = time_ref_to_dict(reference_time)
-        table = Table({"START": start.to("s"), "STOP": stop.to("s")}, meta=meta)
-        return cls(table)
+        table = Table({"START": np.atleast_1d(start), "STOP": np.atleast_1d(stop)})
+
+        return cls(table, reference_time=reference_time)
 
     @classmethod
-    def read(cls, filename, hdu="GTI"):
+    def read(cls, filename, hdu="GTI", format="gadf"):
```
Comment (bkhelifi) [inline gammapy/data/gti.py:112]: The hdu name "GTI" is for the GADF format... Should we assume that the hdu name is independent of the format?

File: gammapy/data/gti.py [hunk 2]
```diff
@@ -101,10 +118,36 @@ def read(cls, filename, hdu="GTI"):
...
+        with fits.open(str(make_path(filename)), memmap=False) as hdulist:
+            return cls.from_table_hdu(hdulist[hdu], format=format)
+
+    @classmethod
+    def from_table_hdu(cls, table_hdu, format="gadf"):
+        """Read from table HDU.
+
+        Parameters
+        ----------
+        table : `~astropy.io.fits.BinTableHDU`
```
Comment (bkhelifi) [inline gammapy/data/gti.py:134]:
```suggestion
        table_hdu : `~astropy.io.fits.BinTableHDU`
```

**Comment by registerrier** [comment]
Regarding the impact on `Observations.select_time`, the computing time is actually reduced with the new approach. The following code snippet runs in 9s with the current main branch and 4s with the proposed model.
```
from gammapy.data import DataStore
from astropy.time import Time

ds = DataStore.from_dir("$GAMMAPY_DATA/hess-dl3-dr1/")
observations = ds.get_observations()

interval1 = [Time('2004-12-04 22:10:10.184'), Time('2004-12-04 22:18:10.184')]
interval2 = [Time('2004-12-04 22:19:00'), Time('2004-12-04 22:38:00')]
interval3 = [Time('2004-12-04 22:50:00'), Time('2004-12-04 23:01:00')]

%timeit filtered = observations.select_time([interval1, interval2, interval3])
```

File: gammapy/data/gti.py [hunk 3]
```diff
@@ -172,14 +220,22 @@ def time_sum(self):
...
     def time_stop(self):
         """GTI end times (`~astropy.time.Time`)."""
-        met = u.Quantity(self.table["STOP"].astype("float64"), "second")
-        return self.time_ref + met
+        return self.table["STOP"]
+
+    @property
+    def met_start(self):
+        """GTI start time difference with reference time in sec, MET (`~astropy.units.Quantity`)."""
+        return (self.time_start - self.time_ref).to("s")
```
Comment (maxnoe) [inline gammapy/data/gti.py:233]: Why go through `self.time_start` here if this is by definition what is stored in the table?

**Comment by maxnoe** [comment]
Any specific reason to use a table with time columns over two time arrays?

File: gammapy/data/gti.py (hunk 1 above)
Comment (registerrier) [reply gammapy/data/gti.py:112]: I am not sure the hduname is really imposed by the gadf.

**Comment by registerrier** [comment]
> Any specific reason to use a table with time columns over two time arrays?

The main reason would be that it guarantees that start and stop times stay in phase and cannot be sorted independently. A minor one would be that we still have to allow access to the `GTI.table` but we could deprecate it.

**Review by AtreyeeS — COMMENTED** [review]
Thanks @registerrier ! This looks good to me. I suppose the internal format should be defined somewhere - how does the user know the default time scale / the table column names...

**Comment by registerrier** [comment]
I have updated the PR and it is ready. 

- Now the `GTI` are represented internally by a `Table` with `START` and `STOP` columns of type `Time`. This has the advantage that start and stop stay in phase when stacking or reordering GTIs.
- The table is validated on init. For now, no specific scale is imposed. Note that previously, it was mostly `TT`. Do we want to enforce that all `GTIs` are in `TT`?
- The reference time is no longer required and should become de-facto metadata. Since the scheme for meta is not yet defined, I have made it a property to stay backward compatible with `GTI.time_ref`
- I have introduced a default value for the reference time which is defined as a global variable `TIME_REF_DEFAULT`. Ideally, this should be part of a set of global configurable variables. 
- I have introduced `met_start` and `met_stop` methods for backward compatibility  with the previous `GTI.table["START"]`

Do you have time for a review @maxnoe or @kosack ?

File: gammapy/data/gti.py [hunk 4]
```diff
@@ -190,27 +258,29 @@ def time_intervals(self):
...
-        start = Time([_[0] for _ in time_intervals]) - reference_time
-        stop = Time([_[1] for _ in time_intervals]) - reference_time
-        meta = time_ref_to_dict(reference_time)
-        table = Table({"START": start.to("s"), "STOP": stop.to("s")}, meta=meta)
-        return cls(table=table)
+        start = Time([_[0] for _ in time_intervals])
+        stop = Time([_[1] for _ in time_intervals])
+
+        if reference_time is None:
+            reference_time = start.min()
```
Comment (adonath) [inline gammapy/data/gti.py:281]: The behavior is well documented, but maybe we should still use the same `TIME_REF_DEFAULT` instead?

**Review by adonath — APPROVED** [review]
Thanks @registerrier! I have left two minor comments, otherwise looks good to me!

**Review by AtreyeeS — APPROVED** [review]
Thanks @registerrier ! No comments at this stage. It will be nice to try this out

File: gammapy/data/gti.py [hunk 5]
```diff
@@ -101,10 +128,36 @@ def read(cls, filename, hdu="GTI"):
             Filename
         hdu : str
             hdu name. Default GTI.
+        format: str
+            Input format, currently only "gadf" is supported
         """
         filename = make_path(filename)
-        table = Table.read(filename, hdu=hdu)
-        return cls(table)
+        with fits.open(str(make_path(filename)), memmap=False) as hdulist:
```
Comment (maxnoe) [inline gammapy/data/gti.py:135]: Why was this change needed? Everywhere else, we use the simple `Table.read`

Comment (registerrier) [reply gammapy/data/gti.py:135]: To call directly the `GTI.from_table_hdu()` method that actually does the gadf to gammapy format conversion.
Alternatively a staticmethod such as `gadf_table_to_gammapy(table)` could also be used.

Comment (adonath) [reply gammapy/data/gti.py:135]: I guess the question is whether meta data accessed via `table.meta` or `hdu.header`. I have not noticed this before but the pattern of `.from_table_hdu()` is unusual. We typically rely on the meta data being present in `table.meta`.

There is still https://github.com/gammapy/gammapy/blob/main/gammapy/maps/axes.py, but if I remember correctly this is only there for backwards compatibility, so we could actually deprecate it. 

@registerrier I think it is worth doing a follow-up PR to change to `GTI.from_table(..., format="gadf")` and not introduce `.from_table_hdu()`...

Comment (adonath) [reply gammapy/data/gti.py:135]: We could keep `.from_table_hdu()` but all it does is really:

```python3
table = Table.read(table_hdu)
return cls.from_table(table, )
```

Comment (adonath) [reply gammapy/data/gti.py:135]: Not sure it's worth to keep, however it saves one line of code...
